### DISCUSSION TITLE
Inquiry about Integrating SPM into ORM_O01 on HL7 v2.5

### DISCUSSION THREAD

**Discussion by gmehmeti** [nHapiNET/nHapi#493] [Q&A] [answered]
I am currently working on a project that involves the integration of laboratory data, and it has become necessary to include specimen-related information within the ORM_O01 message (I am using .Net C#). 
However, I am relatively new to this specific aspect of HL7 v2.5 and could use some assistance in understanding the steps and best practices for achieving this integration.

Thank you

**Comment by milkshakeuk** [comment]
hi @gmehmeti have a look at these tutorials:
https://saravanansubramanian.com/blog/hl72xnhapicreatemessage/

**Reply by gmehmeti** [reply]
Hi @milkshakeuk,
I have checked before to post here, I have contacted also the Saravanan Subramanian.
Also I have asked ChatGPT! 

```csharp
ORM_O01 ormMessage = new ORM_O01();

// Populate the ORM message header
ORM_O01_PATIENT patient = ormMessage.GetPATIENT();
ORM_O01_ORDER order = ormMessage.GetORDER();

// Create an SPM segment
SPM spmSegment = order.GetORDER_DETAIL().GetORDER_DETAIL_OBSERVATION(0).GetSPM();
```
But this code is not Okay!

**Reply by milkshakeuk** [reply]
I will have a look at your sample code.

**Reply by milkshakeuk** [reply]
@gmehmeti what hl7 version are you wanting to use?

**Reply by gmehmeti** [reply]
@milkshakeuk I have mention in first message, I am using HL7 v2.5!

**Reply by milkshakeuk** [reply]
ah yes you did sorry.

**Reply by milkshakeuk** [reply]
@gmehmeti okay the SPM segment is not available within the `ORM_O01` (General Order) though you can add some specimen information in `OBR-11` (Specimen Action Code - [ID](https://hl7-definition.caristix.com/v2/HL7v2.5/DataTypes/ID)), `OBR-14` (Specimen Received Date/Time - [TS](https://hl7-definition.caristix.com/v2/HL7v2.5/DataTypes/TS)), `OBR-15` (Specimen Source -[SPS](https://hl7-definition.caristix.com/v2/HL7v2.5/DataTypes/SPS)).

The only thing I can think of if you need `SPM` is to use `ORU_R01`.

```csharp
using NHapi.Model.V25.Message;

var ormMessage = new ORM_O01();

// Populate the ORM message header
var patient = ormMessage.PATIENT;
var order = ormMessage.GetORDER();

// Create an OBR segment
var obr = order.ORDER_DETAIL.OBR;
obr.SpecimenActionCode.Value = "ID123";
obr.SpecimenReceivedDateTime.Time.Value = "20231102120000";
// obr.SpecimenSource. - lots of fields here
...
```

**Reply by milkshakeuk** [reply]
@gmehmeti have you had the chance to look at this yet?

**Reply by gmehmeti** [reply]
Hello @milkshakeuk ,
The requirement is to create HL7 as follows:

MSH|^~\&|yuuniq|LYUNIQ|SwiSyn|SynlabCH|20231010202001||ORM^O01|ORMY0000000001|P|2.5|||NE|NE|CH|UTF-8|
PID|1||123456||Bond^James||19740312|M|||Sternmatt 6^^Kriens^^6010^CH||+41795248275^^CP^test@gmail.com^^^^^||||||7561234567890
ORC|NW|7810000001|||CM||^^^20231010201859^^R||20231010202001|||LYUNIQ^yuuniq^^^^^^SENDER^SwiSyn
OBR|1|||MUED^Müdigkeit gross|||||||||||781000000101|
SPM|1|781000000101|||E^EDTA-Blut^|||20231010201859

Here is also SPM segment!?

**Reply by milkshakeuk** [reply]
@gmehmeti  there might be an SPM segment in your message but that is non-standard, are you wanting to create a non-standard (custom) message?

**Reply by gmehmeti** [reply]
@milkshakeuk it is possible by using non-standard to create such message that include SPM? 
How?

**Reply by milkshakeuk** [reply]
you can add non-standard segments sure.

```csharp
var ormMessage = new ORM_O01();

ormMessage.AddNonstandardSegment("SPM");
var spm = ormMessage.GetStructure("SPM");

// add stuff to spm
```

**Reply by gmehmeti** [reply]
But to spm variable, I dont have any method to add field?

**Reply by gmehmeti** [reply]
![Untitled](https://github.com/nHapiNET/nHapi/assets/4329476/14c27eac-d804-4401-b21f-9c9c27cdfd05)

**Reply by milkshakeuk** [reply, marked as answer]
```csharp
var ormMessage = new ORM_O01();

ormMessage.AddNonstandardSegment("SPM");
var spm = ormMessage.GetStructure("SPM") as SPM;

// add stuff to spm
```

how about now?

**Reply by gmehmeti** [reply]
Thank you @milkshakeuk

**Reply by gmehmeti** [reply]
@milkshakeuk sorry for interrupt you again, it is possible to add multiple SPM as non standard segment. So for each OBR one or more SPM!?
